src/browser/screen.js

```diff
@@ -375,7 +725,10 @@ function ScreenAdapter(options, screen_fill_buffer)
 
     function update_scale_graphic()
     {
-        elem_set_scale(graphic_screen, scale_x * base_scale, scale_y * base_scale, false);
+        if(!options.disable_autoscale)
+        {
```

Comment:
FYI, I reverted this change in my cleanup, because it breaks `this.set_scale`, and I was under the impression that you wanted to disable the automatic 2x scaling of small resolutions.

Later I realised, that you want v86 to not touch the width/height style at all, because you're using `object-fit: contain` with `width: 100%` or similar.

For now, I'm fine with merging this as-is, but I'm not 100% happy with this API, and will probably make some breaking changes in the future. In particular:

- The 2x auto-scaling should be disableable (or be moved into the caller)
- The `disable_autoscale` option in this PR should be changeable at runtime. I'm thinking of changing it to `options.scale = 0`, so that `set_scale(0)` disables autoscale and `set_scale(1)` restores it

Note that calling `elem_set_scale` is important to get proper 1x resolution on screens where `window.devicePixelRatio` isn't 1. Otherwise, canvas pixels don't render 1-to-1 as screen pixels.

Reply:
First, I never got around to mention it properly, but there is a bug in V86's autoscale. Secondly, the way it is written prevents embedders to apply their own scaling rules to canvas using CSS styles and classes.

The bug is not obvious: Enter fullscreen mode, switch to a graphical mode with small dimensions (320x200, anything that triggers autoscale), and then directly exit fullscreen mode again: The canvas stays zoomed and does not fall back into its non-fullscreen dimensions. The bug is unrelated to graphical text mode, you can test it at https://copy.sh/v86/ using my image [FreeDOS-20m-demo2.img](https://www.rekoba.de/~cs/v86/demo/img/FreeDOS-20m-demo2.img) (20M), steps to reproduce:

    # first boot up image and enter fullscreen
    cd monkey
    monkey.exe
    # now press ESC to exit fullscreen

Before I added code to disable autoscale, I tried to fix this bug. Adding an event listener for [`fullscreenchange`](https://developer.mozilla.org/en-US/docs/Web/API/Document/fullscreenchange_event) event did not help, I believe it's a case for [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) which I felt overblown. Simply disabling it by opt-out seemed like a good middleground to me.

To the second point: The problem crystallizes in [these two lines](https://github.com/copy/v86/blob/d49cb4c65a7a126e45f25141fad3181d79c49359/src/browser/screen.js#L383-L384) in screen.js. Function elem_set_scale() unconditionally rolls over whatever is in these CSS attributes like a train (the same is true about forcing display attribute [`block`](https://github.com/copy/v86/blob/d49cb4c65a7a126e45f25141fad3181d79c49359/src/browser/screen.js#L132) for visibilty, there are more than a dozen other values besides `block` that one could want to use here, like `flex` or `grid`).

In my personal opinion (no one needs to agree) this destructive forcing can easily be avoided by use of CSS classes (which can be overridden by the embedder) instead of direct write access from code to individual element's CSS style attributes for DOM objects that were not created by V86 (like for example the cursor DIV which is out of scope for embedders).

> The disable_autoscale option in this PR should be changeable at runtime. I'm thinking of changing it to options.scale = 0, so that set_scale(0) disables autoscale and set_scale(1) restores it

Good idea, I'll look into it tomorrow.

> Note that calling `elem_set_scale` is important to get proper 1x resolution on screens where `window.devicePixelRatio` isn't 1. Otherwise, canvas pixels don't render 1-to-1 as screen pixels.

I wasn't aware of this and will include it.

Reply:
I hadn't fully understood elem_set_scale() until now, thanks for the clarifications, except one question regarding this small section:

https://github.com/copy/v86/blob/d49cb4c65a7a126e45f25141fad3181d79c49359/src/browser/screen.js#L418-L423

What is the assumed range of window.devicePixelRatio here?

Anyway, I chose the most simple solution to address this in ea2119f for now, though that's still not really what you were describing. The solution is to disable all aspects of automatic scaling when options.scale is set to 0 (as before with disable_autoscale).

An application that sets options.scale to 0 has the full responsibilty for scaling, but is free in how it implements that. That's my propsal for this.

Reply:
> window.devicePixelRatio

It can be any number bigger or equal to 1, including non-integers. The division looks a bit weird, but it indeed does the right thing for all devicePixelRatios. 

> Anyway, I chose the most simple solution to address this in https://github.com/copy/v86/commit/ea2119f094ecbeb30a1070104db5ef27aa0dc992 for now, though that's still not really what you were describing. The solution is to disable all aspects of automatic scaling when options.scale is set to 0 (as before with disable_autoscale).

Yes, that's what I had in mind.

Reply:
> > window.devicePixelRatio
> 
> It can be any number bigger or equal to 1, including non-integers. The division looks a bit weird, but it indeed does the right thing for all devicePixelRatios.

Some made up numbers to make my point:

- if scale is 10 and device_pixel_ratio is any positive integer, scale remains 10
- if scale is 10 and device_pixel_ratio is 2.5, scale becomes 4

If you don't mind, can you elaborate a bit on why these cases are treated so differently? What browser behaviour is being fought here?

I must admit that device_pixel_ratio and HiDPI screens are still a bit of a mystery to me.

Here's my understanding of device_pixel_ratio for you to correct: its value is a combination of the number P of physical pixels per css pixel (which should always be a positive integer, right?) and some small fractional value Z (between 0 and 1, correct?) stemming from the browser's zoom level, so device_pixel_ratio = P + Z.

And about HiDPI screens: On older "96-DPI" screens (like mine) browsers set P to 1, and on HiDPI screens to 2, 3 or 4.

I'm just trying to get my head around that.

> 
> > Anyway, I chose the most simple solution to address this in [ea2119f](https://github.com/copy/v86/commit/ea2119f094ecbeb30a1070104db5ef27aa0dc992) for now, though that's still not really what you were describing. The solution is to disable all aspects of automatic scaling when options.scale is set to 0 (as before with disable_autoscale).
> 
> Yes, that's what I had in mind.

Great, we'll leave it at that, then.